### PR TITLE
Fix for incompatible encodings: UTF-8 and ASCII-8BIT issue #6

### LINKED ISSUE #6: incompatible encodings: UTF-8 and ASCII-8BIT
hello guys,
i have some issue:

`
Exception in filterworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash. {"exception"=>#<Encoding::CompatibilityError: incompatible encodings: UTF-8 and ASCII-8BIT>, "backtrace"=>["org/jruby/RubyString.java:3064:in `gsub'", "/opt/logstash/vendor/jruby/lib/ruby/1.9/uri/common.rb:331:in `unescape'", "/opt/logstash/vendor/jruby/lib/ruby/1.9/uri/common.rb:649:in `unescape'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-urldecode-2.0.2/lib/logstash/filters/urldecode.rb:50:in `urldecode'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-urldecode-2.0.2/lib/logstash/filters/urldecode.rb:39:in `filter'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/filters/base.rb:151:in `multi_filter'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/filters/base.rb:148:in `multi_filter'", "(eval):1223:in `cond_func_40'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):1220:in `cond_func_40'", "(eval):596:in `filter_func'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:244:in `filterworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:237:in `filterworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:178:in `start_filters'"], :level=>:error}
Encoding::CompatibilityError: incompatible encodings: UTF-8 and ASCII-8BIT
           gsub at org/jruby/RubyString.java:3064
       unescape at /opt/logstash/vendor/jruby/lib/ruby/1.9/uri/common.rb:331
       unescape at /opt/logstash/vendor/jruby/lib/ruby/1.9/uri/common.rb:649
      urldecode at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-urldecode-2.0.2/lib/logstash/filters/urldecode.rb:50
         filter at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-urldecode-2.0.2/lib/logstash/filters/urldecode.rb:39
   multi_filter at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/filters/base.rb:151
           each at org/jruby/RubyArray.java:1613
   multi_filter at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/filters/base.rb:148
   cond_func_40 at (eval):1223
           each at org/jruby/RubyArray.java:1613
   cond_func_40 at (eval):1220
    filter_func at (eval):596
   filterworker at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:244
   filterworker at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:237
  start_filters at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.1-java/lib/logstash/pipeline.rb:178
`

How can i fix this issue?
Is there any way to show broken event (DEBUG for this filter)?

### PR DESCRIPTION
**Pull request by guyboertje**
Fixes #6

### CODE REVIEW

File: lib/logstash/filters/urldecode.rb [hunk 1]
```diff
@@ -14,13 +14,21 @@ class LogStash::Filters::Urldecode < LogStash::Filters::Base
...
   # and `cp1252`
   #
   # This setting is useful if your url decoded string are in `Latin-1` (aka `cp1252`)
   # or in another character set other than `UTF-8`.
   config :charset, :validate => ::Encoding.name_list, :default => "UTF-8"
 
+  UnescapeRe = /((?:%[0-9a-fA-F]{2})+)/
+  UnescapePercent = '%'.freeze
+  UnescapePack = 'H*'.freeze
+  JavaUrlDecoder = java.net.URLDecoder
```
Comment (guyboertje) [inline lib/logstash/filters/urldecode.rb]: My bad don't need this line

**Comment by untergeek** [comment]
Nice workaround! LGTM™

File: lib/logstash/filters/urldecode.rb (hunk 1 above)
Comment (untergeek) [reply lib/logstash/filters/urldecode.rb]: I did wonder why that was there, when there were no calls to it...

Comment (untergeek) [reply lib/logstash/filters/urldecode.rb]: Thought it might have been for future expansion 😄

File: spec/filters/urldecode_spec.rb [hunk 2]
```diff
@@ -56,7 +59,61 @@
...
+        }
+      }
+     CONFIG
+     sample("url" => "/fr/search-results?queryText=Organigramme%20de%20la%20Commission%20europ%C3%A9enne%201998&additionalTextParam=organigramme%20de%20la%20commission%20européenne%201998") do
+       expect(subject.get("url")).to eq "/fr/search-results?queryText=Organigramme de la Commission européenne 1998&additionalTextParam=organigramme de la commission européenne 1998"
+       expect(subject.get("tags")).to be_nil
+     end
+   end
+
+   describe "urldecode should not attempt to decode non RFC 3986 compliant strings and set custom tag" do
```
Comment (guyboertje) [inline spec/filters/urldecode_spec.rb]: this needs a new text

File: spec/filters/urldecode_spec.rb [hunk 3]
```diff
@@ -56,7 +59,61 @@
       }
      CONFIG
      sample("message" => "/a/sa/search?rgu=0;+%C3%BB%D3%D0%D5%D2%B5%BD=;+%B7%A2%CB%CD=") do
-       insist { subject.get("message") } == "/a/sa/search?rgu=0;+û\\xD3\\xD0\\xD5ҵ\\xBD=;+\\xB7\\xA2\\xCB\\xCD="
+       expect(subject.get("message")).to eq "/a/sa/search?rgu=0;+û\\xD3\\xD0\\xD5ҵ\\xBD=;+\\xB7\\xA2\\xCB\\xCD="
+       expect(subject.get("tags")).to be_nil
+     end
+   end
+
+   describe "urldecode should not attempt to decode non RFC 3986 compliant strings" do
```
Comment (guyboertje) [inline spec/filters/urldecode_spec.rb]: needs to change text

File: lib/logstash/filters/urldecode.rb [hunk 4]
```diff
@@ -29,23 +36,40 @@ def register
...
-    # If all_fields is true then try to decode them all
-    if @all_fields
-      event.to_hash.each { |name, value| event.set(name, urldecode(value)) }
-    # Else decode the specified field
-    else
-      event.set(@field, urldecode(event.get(@field)))
+    begin
+      # If all_fields is true then try to decode them all
+      if @all_fields
+        event.to_hash.each do |name, value|
```
Comment (untergeek) [inline lib/logstash/filters/urldecode.rb:42]: Thanks for expanding this and making it a bit easier to read

**Comment by untergeek** [comment]
LGTM™

**Pull request merged**